kernel/packages/shared/atlas/reducer.ts

```diff
@@ -30,68 +41,137 @@ export function atlasReducer(state?: AtlasState, action?: AnyAction) {
-        },
-        sceneNames: {
-          ...state.sceneNames,
-          ...(action.payload as { parcels: string[]; name: string }).parcels.reduce((prev, val) => {
-            return { ...prev, [val]: action.payload.name }
-          }, {})
-        }
-      }
-    case FAILURE_NAME_FROM_SCENE_JSON:
-      return {
-        ...state,
-        requestStatus: {
-          ...state.requestStatus,
-          [action.payload.sceneId]: 'failure'
-        }
-      }
+    case QUERY_DATA_FROM_SCENE_JSON:
+      return reduceFetchDataFromSceneJson(state, (action as QuerySceneData).payload)
+    case SUCCESS_DATA_FROM_SCENE_JSON:
+      return reduceSuccessDataFromSceneJson(state, (action as FetchDataFromSceneJsonSuccess).payload.data)
+    case FAILURE_DATA_FROM_SCENE_JSON:
+      return reduceFailureDataFromSceneJson(state, (action as FetchDataFromSceneJsonFailure).payload.sceneId)
     case MARKET_DATA:
-      const marketAction = action as PayloadAction<typeof MARKET_DATA, MarketData>
-      return {
-        ...state,
-        marketName: {
-          ...state.marketName,
-          ...marketAction.payload.data
-        }
-      }
+      return reduceMarketData(state, action.payload)
     case REPORTED_SCENES_FOR_MINIMAP:
-      return {
-        ...state,
-        lastReportPosition: action.payload.reportPosition ? action.payload.reportPosition : state.lastReportPosition,
-        atlasReducer: {
-          ...state.alreadyReported,
-          ...action.payload.parcels.reduce((prev: Record<string, boolean>, next: string) => {
-            prev[next] = true
-            return prev
-          }, {})
-        }
-      }
+      return reduceReportedScenesForMinimap(state, (action as ReportedScenes).payload)
     case DISTRICT_DATA:
-      return {
-        ...state,
-        districtName: {
-          ...state.districtName,
-          ...zip(action.payload.data, (t: District) => [t.id, t.name])
-        }
-      }
+      return reduceDistrictData(state, action)
   }
   return state
 }
+
+function reduceFetchDataFromSceneJson(state: AtlasState, sceneId: string) {
+  const idToScene = { ...state.idToScene }
+
+  if (!idToScene[sceneId]) {
+    idToScene[sceneId] = { ...MAP_SCENE_DATA_INITIAL_STATE }
+  }
+
+  idToScene[sceneId].requestStatus = 'loading'
+
+  return { ...state, idToScene }
+}
+
+function reduceFailureDataFromSceneJson(state: AtlasState, sceneId: string) {
+  const idToScene = { ...state.idToScene }
+
+  if (!idToScene[sceneId]) {
+    idToScene[sceneId] = { ...MAP_SCENE_DATA_INITIAL_STATE }
+  }
+
+  idToScene[sceneId].requestStatus = 'fail'
+
+  return { ...state, idToScene }
+}
+
+function reduceSuccessDataFromSceneJson(state: AtlasState, landData: ILand) {
+  const tileToScene = { ...state.tileToScene }
+  const idToScene = { ...state.idToScene }
+
+  let mapScene: MapSceneData = { ...state.idToScene[landData.sceneId] }
+
+  // NOTE(Brian): this code is for the case in which market data comes first (most likely always)
```

Comment:
Not a big fan of this explanation. Is there any way we can make this go away, some way the code makes this evident?